===================================================================
--- trunk/global.php	(revision 4429)
+++ trunk/global.php	(revision 4430)
@@ -74,6 +74,7 @@
 	{
 		my_setcookie("mybblang", $mybb->settings['bblanguage']);
 	}
+	$mybb->user['language'] = $mybb->settings['bblanguage'];
 }
 // Cookied language!
 else if($mybb->cookies['mybblang'] && $lang->language_exists($mybb->cookies['mybblang']))
