diff -Naur mybb-original/showthread.php mybb-newpost/showthread.php
--- mybb-original/showthread.php	2009-06-26 07:26:35.000000000 +0200
+++ mybb-newpost/showthread.php	2009-08-08 17:22:43.853835474 +0200
@@ -163,6 +163,11 @@
 		}
 	}
 	
+	if($cutoff && $lastread < $cutoff)
+	{
+		$lastread = $cutoff;
+	}
+
 	// Next, find the proper pid to link to.
 	$options = array(
 		"limit_start" => 0,
@@ -1065,4 +1070,4 @@
 	}
 	return $posts;
 }
-?>
\ No newline at end of file
+?>
